[VPlan] Consolidate logic for narrow to single scalars

artagnon opened the pull request:
Consolidate the logic for narrowing to single scalar recipes is in two different places: narrowToSingleScalarRecipes and legalizeAndOptimizeInductions.

artagnon committed:
[VPlan] Consolidate logic for narrow to single scalars

The logic for narrowing to single scalar recipes is in two different
places: narrowToSingleScalarRecipes and legalizeAndOptimizeInductions.
Consolidate them, with minor test changes.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1238,21 +1214,19 @@ static void narrowToSingleScalarRecipes(VPlan &Plan) {
+      // that don't have only their scalar results used (would introduce extra
+      // broadcasts in this case).
+      if (!vputils::isSingleScalar(Def) || Def->getNumUsers() == 0 ||
+          any_of(Def->users(),
+                 [Def](VPUser *U) { return !U->usesScalars(Def); }))
         continue;
 
-      auto *Clone = new VPReplicateRecipe(RepOrWidenR->getUnderlyingInstr(),
-                                          RepOrWidenR->operands(),
-                                          true /*IsSingleScalar*/);
-      Clone->insertBefore(RepOrWidenR);
-      RepOrWidenR->replaceAllUsesWith(Clone);
+      auto *Clone = new VPReplicateRecipe(
+          Def->getUnderlyingInstr(), Def->operands(), true /*IsSingleScalar*/);
+      Clone->insertBefore(Def);
+      Def->replaceAllUsesWith(Clone);
```
Why change the code here? The original name `RepOrWidenR` seems more accurate?

artagnon commented:
Gentle ping.

lukel97 commented on llvm/test/Transforms/LoopVectorize/uniform_across_vf_induction1.ll:
```diff
@@ -57,11 +57,11 @@ define void @ld_div2_step1_start0_ind1(ptr noalias %A, ptr noalias %B) {
-; CHECK-NEXT:    [[BROADCAST_SPLATINSERT:%.*]] = insertelement <2 x i64> poison, i64 [[TMP3]], i64 0
+; CHECK-NEXT:    [[BROADCAST_SPLATINSERT:%.*]] = insertelement <2 x i64> poison, i64 [[TMP2]], i64 0
 ; CHECK-NEXT:    [[BROADCAST_SPLAT:%.*]] = shufflevector <2 x i64> [[BROADCAST_SPLATINSERT]], <2 x i64> poison, <2 x i32> zeroinitializer
+; CHECK-NEXT:    [[TMP3:%.*]] = add nsw <2 x i64> [[BROADCAST_SPLAT]], splat (i64 42)
```
Is this a regression? It looks like we're no longer narrowing this add of the load

artagnon replied:
I investigated further this morning, and I think the Right Thing to do in this case is to unfortunately live with the regression. The hack can never be more than a hack.

artagnon commented:
Gentle ping.

artagnon committed:
[VPlan] Allow wide-stores in narrow to single scalar

Follow up on 132bacd ([VPlan] Also allow extracts as users when
converting to single scalars.) to also allow WidenStore recipes as users
when converting to single scalars. While at it, refine the
vputils::isSingleScalar check to also check vputils::onlyFirstLaneUsed,
and forbid zero users.

The motivation for this patch is to ease consolidation of the logic to
narrow to single scalar, and the consolidation will be tackled in a
follow-up.
